service/windows/src/config.rs

```diff
@@ -0,0 +1,327 @@
+use crate::node::{LogFileConfig, LoggerConfig::*, NodeConfig};
+use anyhow::{anyhow, Context};
+use log::Level;
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use std::fs;
+use std::path::{Path, PathBuf};
+use std::str::FromStr;
+use toml::Value;
+use winreg::{enums::HKEY_LOCAL_MACHINE, RegKey};
+
+/// Look up the config file path in the registry.
+pub fn get_config_file_path() -> anyhow::Result<PathBuf> {
+    let hklm = RegKey::predef(HKEY_LOCAL_MACHINE);
+    let key_name = r"SOFTWARE\Concordium\Node Runner";
+    let node_runner_key = hklm.open_subkey(key_name).with_context(|| {
+        format!(
+            "Could not resolve registry key 'HKEY_LOCAL_MACHINE\\{}'",
+            key_name
+        )
+    })?;
+    let value_name = "Config";
+    let conf: String = node_runner_key.get_value(value_name).with_context(|| {
+        format!(
+            "Could not get registry value 'HKEY_LOCAL_MACHINE\\{}\\{}'",
+            key_name, value_name
+        )
+    })?;
+    Ok(PathBuf::from(conf))
+}
+
+/// Configuration of nodes to run.
+pub struct Config {
+    pub nodes: Vec<NodeConfig>,
+}
+
+/// Convenience macro for looking up a value in a TOML table.
+macro_rules! toml_get_as {
```

Comment:
Is it really necessary to use a macro? Please try to write a normal function, if that is not possible at the very least explain the need for it.

Additionally,  why make a closure and immediately apply it?

Reply:
A macro is more convenient here because it can be variadic. The closure and application allows us to use `?` in `Option`. The macro can be applied to both the `Table` and `Value` types (which have similar `get` functions). The macro can also be called with functions that return either a value or a borrowed pointer.

I've spent a while trying to rewrite this as a function, but I don't think it can be done with a single function that works for all use cases of the macro. I don't think there's enough to be gained to justify putting more effort into rewriting this.

Reply:
It is more convenient to write perhaps, but it is harder to understand. 

I feel that the parsing logic is already hard to follow, and using this macro does not help. Having clearer named type-specific functions would in my view help.